### PR TITLE
Fix/139 correct station disruption processing model

### LINKED ISSUE #139: Fix station disruption processing: DisruptedPoint vs Disruption data structure mismatch
## Problem

The station disruption processing implementation in `fetch_station_disruptions()` is **fundamentally broken** due to a mismatch between expected and actual data structures from the TfL API.

### Current (Broken) Implementation

The code expects:
- **`Disruption`** objects with an `affectedStops` list
- Processing pattern: iterate through disruptions, then iterate through stops within each disruption
- Access fields like `disruption.category`, `disruption.description`, `disruption.affectedStops`

### Actual API Response

The TfL API `StopPoint.DisruptionByMode` endpoint currently returns:
- **`list[DisruptedPoint]`** - each object IS a single station disruption
- Fields: `atcoCode`, `stationAtcoCode`, `description`, `type`, `commonName`, `fromDate`, `toDate`, `appearance`
- No `affectedStops` field - the DisruptedPoint itself represents the affected station

### Evidence

See actual API responses in:
- `backend/tfl_responses/StopPoint_DisruptionByModeByPathModesQueryIncludeRouteBlockedStops_includeRouteBlockedStops_true.txt`
- `backend/tfl_responses/StopPoint_DisruptionByModeByPathModesQueryIncludeRouteBlockedStops_includeRouteBlockedStops_false.txt`

Example from actual response:
```json
{
    "$type": "Tfl.Api.Presentation.Entities.DisruptedPoint, Tfl.Api.Presentation.Entities",
    "atcoCode": "940GZZLUKWG",
    "fromDate": "2024-11-21T13:25:00Z",
    "toDate": "2026-12-23T01:29:00Z",
    "description": "KEW GARDENS STATION: Step-free platform interchange...",
    "commonName": "Kew Gardens Underground Station",
    "type": "Information",
    "mode": "tube",
    "stationAtcoCode": "940GZZLUKWG",
    "appearance": "Information"
}
```

### Current State (Issue #137)

As part of issue #137, the implementation has been **disabled** and returns an empty list to prevent runtime errors:
- Types corrected to use `list[DisruptedPoint]`
- Processing logic commented out
- Warning logged when called
- See `app/services/tfl_service.py` line ~1435

### Step 1: Investigate the Purpose of This Function

**Before making any implementation decisions**, understand the business requirements:

- **What is `fetch_station_disruptions()` for?**
  - Where is it called from in the application?
  - What user-facing features depend on it?
  - What problem does it solve?

- **What data does it need to provide?**
  - Review the `StationDisruptionResponse` schema - what fields are required?
  - Review the `StationDisruption` database model - what gets stored?
  - How is this data displayed to users?
  - What filtering/processing is needed?

- **How is this data used elsewhere?**
  - Are there API endpoints that return this data?
  - Are there frontend components that display it?
  - Are there tests that show expected behavior?

**Action**: Search the codebase for `fetch_station_disruptions`, `StationDisruption`, `StationDisruptionResponse` to understand usage patterns.

### Step 2: Investigate the Correct pydantic-tfl-api Endpoint

**The pydantic-tfl-api library is strongly typed** - types are autogenerated from TfL API OpenAPI specs and highly trustworthy.

- **Find the right endpoint for our needs**:
  - Review the library documentation/source (check `backend/.venv/lib/python3.14/site-packages/pydantic_tfl_api`)
  - Examine `endpoints/*_config.py` files that map OpenAPI queries to response models. Lineclient_config and StoppointClient_config are strong candidates.
  - Look at available `StopPointClient` methods
  - Check if there's a better endpoint than `DisruptionByMode` for our use case

- **Questions to answer**:
  - Is `DisruptionByMode` the right endpoint, or should we use a different one?
  - Does the TfL API have an endpoint that returns `Disruption` objects with `affectedStops`?
  - Should we be using multiple endpoints and combining data?
  - What do the `includeRouteBlockedStops` parameter values change?

- **Compare endpoint options**:
  - `StopPoint.DisruptionByMode` - returns `DisruptedPoint` (what we currently use)
  - Are there other `StopPoint` or `Line` endpoints that might be more appropriate?
  - What's the semantic difference between different disruption endpoints?

**Action**: Thoroughly review pydantic-tfl-api library structure and TfL API documentation to find the optimal endpoint(s).

### Step 3: Map Data Structure to Requirements

Once you understand WHAT we need and WHICH endpoint provides it:

- **If DisruptedPoint is correct**:
  - Map DisruptedPoint fields to StationDisruption model fields
  - Map DisruptedPoint fields to StationDisruptionResponse schema fields
  - Determine how to handle `type` → `disruption_category`
  - Determine how to handle `appearance` → `severity`
  - Handle station lookup via `stationAtcoCode` or `atcoCode`
  - Parse `fromDate`/`toDate` for time periods

- **If we need different endpoint(s)**:
  - Document why DisruptionByMode is insufficient
  - Design new data flow
  - Update function signature if needed

### Step 4: Implement the Solution

Only after completing steps 1-3:

- **Update `_process_station_disruption_data()`**:
  - Remove commented-out code
  - Implement correct processing based on findings from steps 1-3
  - Handle the data structure returned by chosen endpoint(s)

- **Update or replace helper methods**:
  - `_create_station_disruption()` expects `Disruption` - update signature or create new helper
  - Ensure database writes are correct
  - Ensure response objects match schema

- **Handle edge cases**:
  - Stations not in database (skip or log?)
  - Missing required fields
  - Date/time parsing
  - Type/appearance/category mapping

### Step 5: Fix Tests

- **Update test data**:
  - Replace `MockStationDisruption` with correct mock structure
  - Use actual field names from chosen endpoint's response type
  - Match pydantic-tfl-api types exactly

- **Update test assertions**:
  - Verify correct data flow
  - Test edge cases
  - Maintain 95%+ coverage

### Files to Review/Modify

**For Investigation (Steps 1-2)**:
- `backend/app/api/` - Find API endpoints using this function
- `backend/app/schemas/tfl.py` - Review StationDisruptionResponse
- `backend/app/models/tfl.py` - Review StationDisruption model
- `.venv/lib/python3.14/site-packages/pydantic_tfl_api/` - Review library structure
- `backend/tfl_responses/` - Review actual API responses

**For Implementation (Steps 3-5)**:
- `backend/app/services/tfl_service.py`:
  - `_process_station_disruption_data()` (currently disabled at line ~1435)
  - `_create_station_disruption()` (line ~1361) - may need rewrite
  - `fetch_station_disruptions()` (line ~1486)
  
- `backend/tests/test_tfl_service.py`:
  - All station disruption tests (currently skipped)
  - MockStationDisruption class (line ~344)

## Acceptance Criteria

- [ ] **Step 1 complete**: Documented purpose and requirements of station disruption feature
- [ ] **Step 2 complete**: Identified correct pydantic-tfl-api endpoint(s) to use
- [ ] **Step 3 complete**: Data mapping documented and validated
- [ ] Station disruptions correctly processed from chosen endpoint
- [ ] All fields properly mapped to database and response models
- [ ] Station lookup working correctly
- [ ] All edge cases handled
- [ ] All tests updated with correct data structures
- [ ] All tests passing (no skipped tests)
- [ ] Test coverage maintained at 95%+
- [ ] Mypy passes in strict mode
- [ ] Pre-commit hooks pass

## References

- TfL API docs: https://api.tfl.gov.uk/
- pydantic-tfl-api: https://github.com/mnbf9rca/pydantic_tfl_api
- pydantic-tfl-api in venv: `.venv/lib/python3.14/site-packages/pydantic_tfl_api/`
- Related: Issue #137 (type safety improvements)

## Priority

**Medium-High** - This feature is currently completely non-functional and returns empty results.

## Notes for Implementation

- **Do NOT assume** the current endpoint is correct - verify it meets requirements
- **Do NOT guess** at data structures - use actual library types
- The pydantic-tfl-api types are autogenerated from OpenAPI specs and are authoritative
- Check both `includeRouteBlockedStops=true` and `false` responses to understand the difference
- Consider whether we need both types of data or just one

### PR DESCRIPTION
**Pull request by mnbf9rca**
resolves #139

## Summary by Sourcery

Align station disruption processing with the pydantic-tfl-api DisruptedPoint model by adding helper functions, updating database schema and ORM mappings (renaming fields and adding end_date), implementing actual disruption creation logic, and updating tests accordingly.

New Features:
- Add pure helper functions for generating stable disruption IDs, parsing timestamps, and extracting ATCO codes

Enhancements:
- Implement station disruption processing to use DisruptedPoint objects and remove broken placeholder logic
- Rename station disruption model fields (disruption_category→type, severity→appearance) and add end_date
- Update service methods (_create_station_disruption_from_point and _process_station_disruption_data) to map DisruptedPoint fields

Build:
- Add Alembic migration to rename columns and add end_date to station_disruptions table

Tests:
- Replace MockStationDisruption in tests with DisruptedPoint and add coverage for helper functions

### CODE REVIEW

**Comment by mnbf9rca** [comment]
## PR Review Comment Resolution Plan - ✅ COMPLETE

I've analyzed all 8 review comments and addressed them. Here's the final status:

### ✅ **ACCEPTED & FIXED** (7 comments)

1. **Typo fix** (schemas/tfl.py:10): 'pacakge' → 'package' ✅
2. **Test data** (test_tfl_service.py:3077): Fix tfl_id length (15→16 chars) ✅
3. **Docstring** (tfl_service.py:132): Fix invalid hex example ✅
4. **Logging** (tfl_service.py:1497): Add warning when falling back to current time ✅
5. **Bug fix** (tfl_service.py:1501): Handle missing ATCO codes with fallback to station.tfl_id ✅
6. **Test improvement** (test_tfl_service.py:3354): Assert specific hash value (66c2012ca97f6293) ✅
7. **Code quality** (tfl_service.py:194): Revert walrus operator (caused mypy type issue) ✅

### ❌ **REJECTED** (1 comment)

**Comment 6** (tfl_service.py:1522): Wrap db.flush/refresh in try/except
- **Rationale**: Database errors handled at transaction boundaries. Suggested code imports logging in except block (poor practice), catches all exceptions just to log and re-raise (no value). Violates KISS principle. [Responded](https://github.com/mnbf9rca/IsTheTubeRunning/pull/141#discussion_r2526944368)

---

### Test Results ✅

- **All tests passing**: 211/211 ✅
- **Coverage**: 97.73% (exceeds 95% requirement) ✅  
- **Pre-commit hooks**: All passed ✅
  - ruff, mypy, prettier, eslint, tsc all clean
- **New test added**: `test_create_station_disruption_missing_atco_code`

---

### Additional Issue (User)

**'type' field name**: Deferred to separate issue (intentionally matches TfL API, no technical issues)

---

### Summary of Changes

**Files Modified**: 3
- `backend/app/schemas/tfl.py` - Typo fix
- `backend/app/services/tfl_service.py` - Logging, ATCO handling, docstring fix
- `backend/tests/test_tfl_service.py` - Test data fix, hash assertion, new test

**Lines Changed**: ~40 (net addition of 1 test case + logging/fallback logic)

All changes maintain 100% backward compatibility. Ready for merge.

File: backend/app/services/tfl_service.py [hunk 1]
```diff
@@ -1348,49 +1459,79 @@ async def fetch_line_disruptions(
...
-            severity=severity,
+            appearance=appearance,
             tfl_id=tfl_id,
             created_at_source=created_at_source,
+            end_date=end_date,
         )
+
         self.db.add(station_disruption)
+        await self.db.flush()
+        await self.db.refresh(station_disruption)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:1538]: We handle database exceptions at the transaction layer. Adding exception handling here just to log and re-raise adds unnecessary complexity without clear benefit. The suggested code imports logging in an except block (poor practice) and catches all exceptions just to re-raise (no value added). If specific exceptions need handling (e.g., IntegrityError), we address those explicitly where needed. Following KISS principle for this hobby project.

File: backend/app/schemas/tfl.py [hunk 2]
```diff
@@ -6,6 +6,12 @@
 
 from pydantic import BaseModel, ConfigDict, Field
 
+## Note: These are application internal schemas, not direct TfL API schemas.
+## TfL API schemas are sourced from the pydantic-tfl-api pacakge and as
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/schemas/tfl.py]: Fixed in latest commit - corrected typo 'pacakge' → 'package'

File: backend/tests/test_tfl_service.py [hunk 3]
```diff
@@ -2939,33 +3039,44 @@ async def test_fetch_station_disruptions(
...
             station_id=uuid.uuid4(),
             station_tfl_id="940GZZLUKSX",
             station_name="King's Cross",
-            disruption_category="RealTime",
+            type="Interchange Message",
             description="Lift out of service",
-            severity="Lift Closure",
-            tfl_id="disruption-1",
+            appearance="RealTime",
+            tfl_id="abc123def456789",
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_tfl_service.py]: Fixed in latest commit - changed test tfl_id from 15 to 16 characters: 'abc123def4567890'

File: backend/app/services/tfl_service.py [hunk 4]
```diff
@@ -88,6 +96,109 @@
...
+
+    Example:
+        >>> from datetime import datetime, UTC
+        >>> _generate_station_disruption_tfl_id(
+        ...     "940GZZLUEUS",
+        ...     datetime(2025, 11, 14, 10, 0, tzinfo=UTC),
+        ...     datetime(2025, 11, 15, 18, 0, tzinfo=UTC),
+        ...     "Lift out of service"
+        ... )
+        'a1b2c3d4e5f6g7h8'
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: Fixed in latest commit - corrected docstring example to use valid hex: 'a1b2c3d4e5f67890'

File: backend/app/services/tfl_service.py [hunk 5]
```diff
@@ -1348,49 +1459,79 @@ async def fetch_line_disruptions(
...
+        appearance: str | None = getattr(disrupted_point, "appearance", None)
+        # Handle None or missing description (field must not be NULL in DB)
+        description: str = getattr(disrupted_point, "description", None) or "No description available"
+
+        # Parse timestamps using pure helper function
+        from_date_str: str | None = getattr(disrupted_point, "fromDate", None)
+        to_date_str: str | None = getattr(disrupted_point, "toDate", None)
+
+        created_at_source = _parse_tfl_timestamp(from_date_str) or datetime.now(UTC)
+        end_date = _parse_tfl_timestamp(to_date_str)
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:1505]: Fixed in latest commit - added logger.warning when fromDate is missing and we fall back to current time. Uses structured logging with context (from_date_str, truncated disrupted_point data).

File: backend/app/services/tfl_service.py [hunk 6]
```diff
@@ -1348,49 +1459,79 @@ async def fetch_line_disruptions(
...
+
+        # Generate TfL ID using pure helper function
+        station_atco = _extract_station_atco_code(disrupted_point) or ""
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py]: Fixed in latest commit - now checks if ATCO code is missing and logs a warning. Falls back to station.tfl_id (already validated by caller) instead of empty string. This prevents non-unique hash IDs. Added test: test_create_station_disruption_missing_atco_code.

File: backend/tests/test_tfl_service.py [hunk 7]
```diff
@@ -3207,38 +3316,43 @@ async def test_create_station_disruption(db_session: AsyncSession) -> None:
...
     # Verify database record was created
     await db_session.commit()
-    result_db = await db_session.execute(select(StationDisruption).where(StationDisruption.tfl_id == "disruption-123"))
+    result_db = await db_session.execute(select(StationDisruption).where(StationDisruption.tfl_id == result.tfl_id))
     disruption = result_db.scalar_one_or_none()
     assert disruption is not None
     assert disruption.station_id == station.id
-    assert disruption.disruption_category == "PlannedWork"
+    assert disruption.type == "Information"
+    assert disruption.appearance == "PlannedWork"
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/tests/test_tfl_service.py:3356]: Fixed in latest commit - added assertion for exact hash value (66c2012ca97f6293) to verify deterministic hash function behavior and catch unintended changes.

File: backend/app/services/tfl_service.py [hunk 8]
```diff
@@ -88,6 +96,109 @@
...
+        ATCO code string (e.g., '940GZZLUEUS'), or None if both fields missing
+
+    Example:
+        >>> point = DisruptedPoint(stationAtcoCode="940GZZLUEUS", ...)
+        >>> _extract_station_atco_code(point)
+        '940GZZLUEUS'
+    """
+    # stationAtcoCode is more explicit, use it if available
+    station_code: str | None = getattr(disrupted_point, "stationAtcoCode", None)
+    if station_code:
```
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:194]: Cannot apply - walrus operator causes mypy strict mode error: 'Returning Any from function declared to return str | None'. Explicit type annotation (station_code: str | None = getattr(...)) is required for type safety. Kept original pattern.

File: backend/app/services/tfl_service.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:1538]: Rejected: We handle database exceptions at the transaction layer. Adding exception handling here just to log and re-raise adds unnecessary complexity. See rejection rationale in previous comment.

File: backend/app/services/tfl_service.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mnbf9rca) [reply backend/app/services/tfl_service.py:194]: Rejected: Walrus operator causes mypy strict mode error ('Returning Any from function'). Explicit type annotation required for type safety. See rejection rationale in previous comment.
